lib/AST/RequirementMachine/RuleBuilder.cpp

```diff
@@ -380,6 +381,16 @@ void RuleBuilder::addRequirement(const Requirement &req,
                    : Context.getSubstitutionSchemaFromType(
                         otherType, proto, result));
 
+      // If 'T' is a parameter pack, this is a same-element
+      // requirement that becomes the following rewrite rule:
+      //
+      //   [element].T.[concrete: C<X, Y>] => [element].T
+      if (subjectType->isParameterPack()) {
```

Comment:
But only if the constraint type does not contain any packs right?

Reply:
Yeah, that's right. I'm not sure what the rewrite rule looks like if the constraint type contains packs but the subject type does not? That's still conceptually a same-element requirement.